src/mesh/mesh.cpp

```diff
@@ -1042,15 +1044,34 @@ void Mesh::Initialize(bool init_problem, ParameterInput *pin, ApplicationInput *
     for (int i = 0; i < nmb; ++i) {
       block_list[i]->meshblock_data.Get()->StartReceiving(BoundaryCommSubset::mesh_init);
     }
-    // send conserved variables
-    for (int i = 0; i < nmb; ++i) {
-      block_list[i]->meshblock_data.Get()->SendBoundaryBuffers();
+
+    const int num_partitions = DefaultNumPartitions();
+    // send FillGhost variables
+    for (int i = 0; i < num_partitions; i++) {
+      auto &md = mesh_data.GetOrAdd("base", i);
+      cell_centered_bvars::SendBoundaryBuffers(md);
     }
 
-    // wait to receive conserved variables
-    for (int i = 0; i < nmb; ++i) {
-      block_list[i]->meshblock_data.Get()->ReceiveAndSetBoundariesWithWait();
+    // wait to receive FillGhost variables
+    // TODO(someone) evaluate if ReceiveWithWait kind of logic is better, also related to
+    // https://github.com/lanl/parthenon/issues/418
+    bool all_received = true;
+    do {
+      all_received = true;
+      for (int i = 0; i < num_partitions; i++) {
+        auto &md = mesh_data.GetOrAdd("base", i);
+        if (cell_centered_bvars::ReceiveBoundaryBuffers(md) != TaskStatus::complete) {
+          all_received = false;
+        }
+      }
+    } while (!all_received);
+
+    // unpack FillGhost variables
+    for (int i = 0; i < num_partitions; i++) {
+      auto &md = mesh_data.GetOrAdd("base", i);
+      cell_centered_bvars::SetBoundaries(md);
```

Comment:
This is such a small changeset. Can we cage this behind an ifdef or runtime parameter? That way codes using sparse variables will still run and other codes can benefit.